File: concurrent-browser-sessions/rest/python/concurrent.py
# Captures screenshots of multiple pages concurrently using the Browserless REST API.
# Uses ThreadPoolExecutor so requests run in parallel threads.
#
# Install: pip install requests
# Run:     python concurrent.py

import requests
from concurrent.futures import ThreadPoolExecutor, as_completed

TOKEN = 'YOUR_API_TOKEN_HERE'
URLS = [
    'https://example.com/page/1',
    'https://example.com/page/2',
    'https://example.com/page/3',
    'https://example.com/page/4',
    'https://example.com/page/5',
]


def capture(i_url):
    i, url = i_url
    res = requests.post(
        f'https://production-sfo.browserless.io/screenshot?token={TOKEN}',
        json={'url': url, 'options': {'type': 'png', 'fullPage': True}},
    )
    filename = f'screenshot-{i + 1}.png'
    with open(filename, 'wb') as f:
        f.write(res.content)
    print(f'Saved {filename}')


with ThreadPoolExecutor(max_workers=5) as executor:
    futures = [executor.submit(capture, (i, url)) for i, url in enumerate(URLS)]
    for f in as_completed(futures):
        f.result()

print('All done')

File: concurrent-browser-sessions/rest/nodejs/concurrent.mjs
// Captures screenshots of multiple pages concurrently using the Browserless REST API.
// Requires Node.js 18+ for the native fetch API. No extra packages needed.
//
// Run: node concurrent.mjs

import fs from 'fs';

const TOKEN = 'YOUR_API_TOKEN_HERE';
const URLS = [
  'https://example.com/page/1',
  'https://example.com/page/2',
  'https://example.com/page/3',
  'https://example.com/page/4',
  'https://example.com/page/5',
];

await Promise.all(
  URLS.map((url, i) =>
    fetch(`https://production-sfo.browserless.io/screenshot?token=${TOKEN}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ url, options: { type: 'png', fullPage: true } }),
    }).then(async (res) => {
      const buf = Buffer.from(await res.arrayBuffer());
      fs.writeFileSync(`screenshot-${i + 1}.png`, buf);
      console.log(`Saved screenshot-${i + 1}.png`);
    })
  )
);

console.log('All done');
